### PR TITLE
Concurrent sessions

### PR DESCRIPTION
**Pull request by DanSipola**
Hi!

Just checking if you would be willing to accept a hack like this (totally understand if you don't :). It's work in progress at the moment, I got as far as making it work. I've got some time to fix the tests and perhaps refine it a bit if you want.

So my use case is that I occasionally need to run puppet bolt towards a bunch of nodes. In order to run tunnelled sessions concurrently I needed to tweak the application a bit to avoid race conditions with the static private key file. So I changed from writing the private key to a static path to writing the key per instance id.

Unfortunately it seemed that bolt wasn't able to authenticate with the private key on disk, I don't really know why and didn't investigate further. So I added the option to run ssh and ssh-proxy in "agent" mode which basically just adds the key to the ssh agent (with a short lifetime) and writes the public key to disk instead. Apparently if ssh finds a public key in the IdentityFile it will try to load the corresponding private key from the agent. Seems to work fine.

### CODE REVIEW

**Comment by DanSipola** [comment]
Hi! No problem. I actually found that they've added an option for a custom ssh executable to puppet bolt, so something similar to what you are suggesting works for me. In this PR I implemented it so that generated keys are written to "instanceid.aws_region.aws_profile" and you refer to them in ssh config with "IdentityFile ~/.aws-gate/%h". This way I can start one connection per node simultaneously.

So what if I start by removing the whole ssh agent thing, but keep the change with regards to the path keys are written to?

**Comment by DanSipola** [comment]
I really quickly removed the ssh agent stuff so you can see what I mean. Something like this supports my use case. When you have time, let me know what you think. No rush!

**Comment by xen0l** [comment]
Yeah, that would work for me. Though, some things to consider:
* keys should be placed under a directory in "~/.aws-gate", so if you are using more of them, they don't pollute `~/.aws-gate` and could be easily removed.
*  the default key path should default to something like `~/.aws-gate/keys/default`
* argument should be added to required subcommands. I tried to follow arguments, so they are comparatble to ssh(1), but in this case `-f` is used for putting the ssh in the backgound, but `ssh-keygen` uses `-f` to specify the path. What do  you think about `-i`?

**Comment by DanSipola** [comment]
Sounds good. I have updated the PR with a quick implementation of your suggestions. It only covers the commands ssh-proxy and ssh-config as it doesn't seem relevant for the others (in my opinion they should just work without too much configuration).

I actually used -f for the filename argument as I would expect -i to take an existing file, and since the program attempts to remove the file it may cause unnecessary annoyance if that would happen. I'm happy to change though.

So my use case is supported with an ssh config generated with
```bash
$ aws-gate ssh-config -f %h -l ubuntu -p profile -r region
Host *.region.profile
IdentityFile /home/user/.aws-gate/keys/%h
IdentitiesOnly yes
User ubuntu
Port 22
ProxyCommand sh -c "aws-gate ssh-proxy -l ubuntu -f %h -p `echo %h | sed -Ee 's/^(.*)\.(.*)\.(.*)$/\\3/g'` -r `echo %h | sed -Ee 's/^(.*)\.(.*)\.(.*)$/\\2/g'` `echo %h | sed -Ee 's/^(.*)\.(.*)\.(.*)$/\\1/g'`"
```
If this is getting closer to something that would work I will start to look at the documentation and tests.

File: aws_gate/ssh_config.py [hunk 1]
```diff
@@ -6,31 +6,31 @@
...
 @valid_aws_profile
 @valid_aws_region
 def ssh_config(
     profile_name=AWS_DEFAULT_PROFILE,
     region_name=AWS_DEFAULT_REGION,
     user=DEFAULT_OS_USER,
     port=DEFAULT_SSH_PORT,
+    key_name=DEFAULT_KEY_NAME,
 ):
+    PROXY_COMMAND = [
```
Comment (xen0l) [inline aws_gate/ssh_config.py:25]: Why move this?

**Comment by xen0l** [comment]
@DanSipola yes, looks reasonable good!

**Comment by DanSipola** [comment]
Hi!

Sorry for this taking so long time. I've had a long summer vacation and then I have been trying this out for a while. Turns out that you get rate limited by the aws api rather quick when you run with some concurrency so it doesn't really work the way I wanted. I suppose that you don't want to proceed with this as it only complicates things.

I will close this PR now (and add another one for just passing the environment to subprocesses). Thanks for your attention and sorry for taking up your time!

**Pull request closed without merging**